ibft/ibft_algorithm_test.go

```diff
@@ -0,0 +1,79 @@
+package ibft
+
+import (
+	"testing"
+	"time"
+
+	"github.com/herumi/bls-eth-go-binary/bls"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap/zaptest"
+
+	msgcontinmem "github.com/bloxapp/ssv/ibft/msgcont/inmem"
+	"github.com/bloxapp/ssv/ibft/proto"
+	"github.com/bloxapp/ssv/utils/dataval/bytesval"
+)
+
+// IBFT ALGORITHM 2: Happy flow - a normal case operation
+func TestUponPrePrepareMessagesBroadcastsPrepare(t *testing.T) {
+	secretKeys, nodes := generateNodes(4)
+	instance := setupInstance(t, nodes, secretKeys)
+
+	// Upon receiving valid PRE-PREPARE messages - 1, 2, 3
+	firstMessage := setupMessage(1, secretKeys[1], proto.RoundState_PrePrepare)
+	instance.prePrepareMessages.AddMessage(firstMessage)
+
+	secondMessage := setupMessage(1, secretKeys[2], proto.RoundState_PrePrepare)
+	instance.prePrepareMessages.AddMessage(secondMessage)
+
+	thirdMessage := setupMessage(1, secretKeys[3], proto.RoundState_PrePrepare)
+	instance.prePrepareMessages.AddMessage(thirdMessage)
+
+	require.NoError(t, instance.uponPrePrepareMsg().Run(thirdMessage))
+
+	// ...such that JUSTIFY PREPARE is true
+	runJustifyPrePrepare(t, instance, 1)
+
+	// broadcasts PREPARE message
+	prepareMessage := setupMessage(1, secretKeys[1], proto.RoundState_Prepare)
+	instance.prepareMessages.AddMessage(prepareMessage)
+}
+
+func setupMessage(id uint64, secretKeys *bls.SecretKey, roundState proto.RoundState) *proto.SignedMessage {
```

Comment:
If we agree to segregate data setup and test function logic
i will extract common setup for a "test_utils.go" at _../ibft/ folder_

+  refactor unit tests to re-use them